lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,744 @@
+        setExpr(op.getResult(), e);
+      })
+      .Case<DivPrimOp>([&](auto op) {
+        auto lhs = getExpr(op.lhs());
+        Expr *e;
+        if (op.getType().isSigned()) {
+          e = solver.add(lhs, solver.known(1));
+        } else {
+          e = lhs;
+        }
+        setExpr(op.getResult(), e);
+      })
+      .Case<RemPrimOp>([&](auto op) {
+        auto lhs = getExpr(op.lhs());
+        auto rhs = getExpr(op.rhs());
+        auto *e = solver.min(lhs, rhs);
+        setExpr(op.getResult(), e);
+      })
+      .Case<AndPrimOp, OrPrimOp, XorPrimOp>([&](auto op) {
+        auto lhs = getExpr(op.lhs());
+        auto rhs = getExpr(op.rhs());
+        auto e = solver.max(lhs, rhs);
+        setExpr(op.getResult(), e);
+      })
+      .Case<LEQPrimOp, LTPrimOp, GEQPrimOp, GTPrimOp, EQPrimOp, NEQPrimOp>(
+          [&](auto op) {
+            auto e = solver.known(1);
+            setExpr(op.getResult(), e);
+          })
+      .Case<MuxPrimOp>([&](auto op) {
+        // Caveat: The Scala implementation imposes a constraint on the select
+        // signal to be at least 1 bit wide. The FIRRTL MLIR dialect requries
+        // the select signal to be exactly `uint<1>` anyway, so this constraint
+        // is not needed here.
+        auto high = getExpr(op.high());
+        auto low = getExpr(op.low());
+        auto e = solver.max(high, low);
+        setExpr(op.getResult(), e);
+      })
+      // Handle the various connect statements that imply a type constraint.
+      .Case<ConnectOp>([&](auto op) {
+        auto dest = getExpr(op.dest());
+        auto src = getExpr(op.src());
+        unifyTypes(dest, src);
+      });
+  // TODO: Handle PartialConnect
+  // TODO: Handle DefRegister
+  // TODO: Handle Attach
+  // TODO: Handle Conditionally
+}
+
+/// Declare free variables for the type of a value, and associate the resulting
+/// set of variables with that value.
+Expr *InferenceMapping::declareVars(Value value) {
+  Expr *e = declareVars(value.getType());
+  setExpr(value, e);
+  return e;
+}
+
+/// Declare free variables for a type.
+Expr *InferenceMapping::declareVars(Type type) {
+  if (auto ftype = type.dyn_cast<FIRRTLType>())
+    return declareVars(ftype);
+  return solver.nil();
+}
+
+/// Declare free variables for a FIRRTL type.
+Expr *InferenceMapping::declareVars(FIRRTLType type) {
+  // TODO: Support aggregate and compound types as well.
+  auto width = type.getBitWidthOrSentinel();
+  if (width >= 0) {
+    return solver.known(width);
+  } else if (width == -1) {
+    return solver.var();
+  } else if (auto inner = type.dyn_cast<FlipType>()) {
+    return declareVars(inner.getElementType());
+  } else {
+    return solver.nil();
+  }
+}
+
+/// Unifies the types `larger` and `smaller` and establishes constraints to
+/// ensure the sizes in the larger one are greater than or equal to the sizes in
+/// the smaller one.
+void InferenceMapping::unifyTypes(Expr *larger, Expr *smaller) {
```

Comment:
Yeah probably not the best of names. I was inclined to use "unify" here because later on when we have aggregate types, this looks a lot like [unification during type inference](https://www.cs.cornell.edu/courses/cs3110/2011sp/Lectures/lec26-type-inference/type-inference.htm). But maybe `addGEQConstraint` is a better name here :wink: